### PR TITLE
fix: fix custom color displaying wrong classname & alignment overflow

### LINKED ISSUE #39386: Timeline right has some problems with its style & classname error
### Reproduction link

[Edit on CodeSandbox](https://codesandbox.io/s/ji-ben-yong-fa-antd-5-0-4-forked-3iq0c8?file=/demo.tsx)

### Steps to reproduce

https://user-images.githubusercontent.com/32004925/206284633-f1ddb407-a2db-4be4-872a-eed17701ce67.mp4

### What is expected?

No overflow container

Do not show classname when customizing styles

### What is actually happening?

Right alignment is out of the container

| Environment | Info |
| --- | --- |
| antd | 5.0.4 |
| React | any |
| System | macOS 12.5 |
| Browser | Chrome: 107.0.5304.110 |

### PR DESCRIPTION
**Pull request by Wxh16144**
[[中文版模板 / Chinese template](https://github.com/ant-design/ant-design/blob/master/.github/PULL_REQUEST_TEMPLATE/pr_cn.md)]

### 🤔 This is a ...

- [ ] New feature
- [x] Bug fix
- [ ] Site / documentation update
- [ ] Demo update
- [x] Component style update
- [x] TypeScript definition update
- [ ] Bundle size optimization
- [ ] Performance optimization
- [ ] Enhancement feature
- [ ] Internationalization
- [ ] Refactoring
- [ ] Code style optimization
- [ ] Test Case
- [ ] Branch merge
- [ ] Workflow
- [ ] Other (about what?)

### 🔗 Related issue link

fix #39386

### 📝 Changelog

| Language   | Changelog |
| ---------- | --------- |
| 🇺🇸 English |     fix custom color displaying wrong classname & alignment overflow      |
| 🇨🇳 Chinese |     修复自定义颜色显示错误类名和对齐溢出的问题      |

### ☑️ Self-Check before Merge

⚠️ Please check all items below before requesting a reviewing. ⚠️

- [x] Doc is updated/provided or not needed
- [x] Demo is updated/provided or not needed
- [x] TypeScript definition is updated/provided or not needed
- [x] Changelog is provided or not needed

### CODE REVIEW

File: components/timeline/TimelineItem.tsx [hunk 1]
```diff
@@ -45,10 +55,10 @@ const TimelineItem: React.FC<TimelineItemProps> = ({
   const dotClassName = classNames({
     [`${prefixCls}-item-head`]: true,
     [`${prefixCls}-item-head-custom`]: !!dot,
-    [`${prefixCls}-item-head-${color}`]: true,
+    [`${prefixCls}-item-head-${color}`]: PresetProcessColorRegex.test(color),
```
Comment (MadCcc) [inline components/timeline/TimelineItem.tsx:58]: 用 presetColor 和 status 是不就够了

Comment (Wxh16144) [reply components/timeline/TimelineItem.tsx:58]: 没理解呢

presetColor 里面没有 gray;

这里我是看其他的预设值都在 `_util/colors.ts` 中，所以我把这里的也顺便移动过去了，但是如何命名可能有些困难。

Comment (MadCcc) [reply components/timeline/TimelineItem.tsx:58]: 这个居然没有支持全预设颜色
image

Comment (Wxh16144) [reply components/timeline/TimelineItem.tsx:58]: 那搞个计划，在 common 里面统一成一个？

另外我发现 color 的 TS 类型推导没有生效，我前面考虑的是写成这样的。

image

Comment (Wxh16144) [reply components/timeline/TimelineItem.tsx:58]: > 这个居然没有支持全预设颜色 image

要不这个在 feature 分支合并后再修改吧

**Comment by afc163** [comment]
冲突了

**Comment by MadCcc** [comment]
这个 PR 先挂起来吧，等 https://github.com/ant-design/ant-design/pull/39742

**Review by afc163 — APPROVED** [review]

**Comment by MadCcc** [comment]
Ready 了就把 WIP 去掉？

**Review by zombieJ — APPROVED** [review]

**Pull request merged**